File system and Binary file sending

ionite34 opened the pull request:
Closes #114, #103

## Implementing a memory-based file system, running written python files, and sending binary files from snekbox to requester.

## Permission specifications
1. We will create a container directory at `/snekbox/memfs`, which will be mounted as a tmpfs memory directory. The `memfs` folder is execute-only for user code with no read/write permissions.
2. Each snekbox run instance will have a new mounted tmpfs on the docker system created with a unique ID, a subfolder of this (`/home`) is then mounted as R/W folders onto the NSJail instance.
3. No instance should be able to read other instance filesystem mounts as they are not mounted to NSJail. 
```yaml
# Docker host
|- /memfs/ab043b4e-5955-4734-b8e2-998c0f6c7131/home/main.py

# Visible to instance user
|- /home/main.py
```
4. This UUID directory is read/execute only for the user code
5. There is also a `home` folder, which has read/write permission for nsjail, and is mounted into nsjail as a R/W folder at `/home`
6. The `$HOME` env var has also been set to point to this mounted folder in the nsjail. (Required for libraries like matplotlib to work)

## Usage
The user has a cwd in `/home` which they can read/write in. In this directory there will be a `output` folder. Any files written within this folder will be uploaded.

The file system per instance is default to be limited at 48MiB, can be changed in the cfg file.

## Current plan

### File system
- [x] 1. Implement a unique tempdir generator backed by `dev/shm`
- [x] 2. Change runtime config to mount an unique tempdir to each snekbox execution
- [x] 3. Make this the working directory for the python file
- [x] 4. Use a written python file for better compatibility with file usage (`__file__`) and other usages such as `inspect.getsource`

### Sending binary files via API
- [x] 1. Scan for created files after snekbox execution completes
- [x] 2. Serialize binary file into JSON, with data and name (possible base64 encoding and / or compression)
- [x] 3. Add API route
- [x] 4. Add a user-code side method for files to be marked sent, or some standard (like naming the file "out.*")
- [x] 5. Formalize API route specifications and docs

## API Request changes
> Backwards-compatible usage of `input` is retained. If `input` is supplied, `args` defaults to `-c`. Either one of `input` or `args` must be supplied.

Compatible mode with input
```py
{
    "input": "[i for i in range(1000)]",
    "args": ["-m", "timeit"]
}
```
New args without input
```py
{
    "args": ["-m", "timeit", "[i for i in range(1000)]"]
}
```
Sending files to be run
```py
{
    "args": ["main.py"],
    "files": [
        {
            "name": "main.py",
            "content": "print(1)"
        }
    ]
}
```

## API response additions

(Previous)
```py
{
	"stdout": "Hello\n",
	"returncode": 0
}
```

(New)
New dictionary key `files` of type `list[dict]`, holding dictionaries representing attachments
Sub-dictionaries will have type `dict[str, str | int]`
Attachment dict keys:
- "name" (str)
- "size" (int) Content size in bytes (before compression)
- "content" (str) Base64 encoded bytes
```py
{
	"stdout": "Hello\n",
	"returncode": 0,
	"files": [
		{
			"name": "output.png",
			"size": 57344,
			"content": "SGVsbG8...=",
		},
	]
}
```

## Demo Images

![image](https://user-images.githubusercontent.com/13956642/202094964-ecdfb6ee-e983-4855-9ca6-a68e268787b4.png)
![image](https://user-images.githubusercontent.com/13956642/202094977-13305e9f-9968-4489-a00e-0b1111cc8f22.png)

<img width="559" alt="image" src="https://user-images.githubusercontent.com/13956642/202094993-79b85712-777d-436d-8001-53c0837206a9.png">

<img width="559" alt="image" src="https://user-images.githubusercontent.com/13956642/202100681-aeb0650b-a28b-47a1-a693-26db8d9c653b.png">

wookie184 commented on snekbox/nsjail.py:
```diff
@@ -152,8 +166,28 @@ def python3(
+        with NamedTemporaryFile() as nsj_log, MemFS(self.memfs_instance_size) as fs:
+            # Add the temp dir to be mounted as cwd
+            nsjail_args = (
+                # Mount a tmpfs at /dev/shm to support multiprocessing
+                "--mount",
+                # src:dst:fs_type:options
+                f"{fs.shm}:/dev/shm:tmpfs:size={fs.instance_size}",
```
I'm struggling to work out exactly what this does, does this create a new tmpfs (I would guess it does?)? If it does, then surely the `src` does nothing? I think the approach in #113 of defining `/dev/shm` in `snekbox.cfg` and just letting nsjail handle it makes more sense.

ionite34 replied:
My understanding (from skimming over the barely existant documentation and confusing cpp parser source, is that this mounts a regular folder (in this case backed by tmpfs on the docker side) onto nsjail as a `tmpfs`, so it isn't actually a tmpfs created on the nsjail side, but a fake one? This might be completely wrong, I'm not sure.

However the space here is usable to the full limit of our memfs allocation, and not the memory limit of the instance, so that would suggest it's not using nsjail memory? 

In any case there seems to be 0 difference in whether this is a bindmount (like `home`) or `tmpfs` mount as far as I can tell? I opted for `tmpfs` as it seems more "realistic" as far as the nsjail/python side is concerned on what a `/dev/shm` should be.

wookie184 replied:
> My understanding (from skimming over the barely existant documentation and confusing cpp parser source), is that this mounts a regular folder (in this case backed by tmpfs on the docker side) onto nsjail as a tmpfs, so it isn't actually a tmpfs created on the nsjail side, but a fake one? This might be completely wrong, I'm not sure.

I did some testing and I think my guess was correct and it just creates a new tmpfs ignoring the source. To demonstrate this you can modify the code to this:
```python
            (fs.shm / "test.txt").write_text("hello world")
            log.info(list(fs.shm.iterdir()))

            try:
                nsjail = subprocess.Popen(
                    args, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, text=True
                )
            except ValueError:
                return EvalResult(args, None, "ValueError: embedded null byte")
            
            log.info(list(fs.shm.iterdir()))
```
And try sending this to snekbox:
```python
import os
open('/dev/shm/thinggggggg.txt', 'w').write('hello world')
print(os.listdir('/dev/shm'))
```
In the logs you'll get something like
```
snekbox_dev  | 2022-11-18 23:23:14,519 |     8 |                 snekbox.nsjail |     INFO | [PosixPath('/memfs/0562f335-93b6-4db6-89d0-1ff580b58921/dev/shm/test.txt')]  
snekbox_dev  | 2022-11-18 23:23:14,520 |     8 |                 snekbox.nsjail |     INFO | [PosixPath('/memfs/0562f335-93b6-4db6-89d0-1ff580b58921/dev/shm/test.txt')] 
```
and in the stdout response from snekbox you'll get
```
['thinggggggg.txt']
```
I also tried replacing `{fs.shm}` with random characters in `f"{fs.shm}:/dev/shm:tmpfs:size={fs.instance_size}"` and it didn't seem to affect how it worked.

> In any case there seems to be 0 difference in whether this is a bindmount (like home) or tmpfs mount as far as I can tell? I opted for tmpfs as it seems more "realistic" as far as the nsjail/python side is concerned on what a /dev/shm should be.

There is a difference (bindmount uses the existing tmpfs), and whilst it doesn't really matter in this case, this current code is confusing (and the code to create `fs.shm` is redundant) so I do think the approach in #113 is better.

ionite34 replied:
Okay cool thanks for the testing and info. Yeah I agree in that case the config approach in #113 is probably better then and will be remove the shm mounting in this PR.

ionite34 replied:
Removed in [9327283](https://github.com/python-discord/snekbox/pull/159/commits/932728308f6773bb525028281ccf1a598d20786b)

wookie184 commented on snekbox/nsjail.py:
```diff
@@ -3,17 +3,21 @@
 
 from google.protobuf import text_format
 
+# noinspection PyProtectedMember
```
out of interest, what is pycharm complaining about to make this necessary?

ionite34 replied:
The fact that `DEBUG` isn't declared in `__all__`, which I guess I understand the intention there?

wookie184 replied:
I think we should probably just add `DEBUG` to `__all__`

ionite34 replied:
Added in [893dbd7](https://github.com/python-discord/snekbox/pull/159/commits/893dbd72d6ba2d5d6dfe979795a94d63c3791b97)

wookie184 commented on snekbox/nsjail.py:
```diff
@@ -163,9 +200,27 @@ def python3(
+            # Filter out empty strings (causes issues with python cli)
+            args.extend(s for s in py_args if s)
+
+            c_arg = "c" in "".join(py_args)
+
+            # Override for `timeit`
+            if "timeit" in py_args:
+                use_file = False
+
+            match (use_file, c_arg):
+                case (True, _) | (None, False):
+                    args.append("main.py")
+                    # Write the code to a file
+                    with fs.allow_write():
+                        code_path = fs.home / "main.py"
+                        code_path.write_text(code)
+                        log.info(f"Created code file at [{code_path!r}].")
+                case _:
+                    args.append(code)
 
```
This feels pretty fragile, changing behaviour just based on whether there is a letter `c` in the arguments and having to hardcode support in for timeit isn't great.

Maybe we should make the `input` API parameter optional, and then only add `main.py` if that was passed, otherwise just use the args given (which could contain code). This would be pretty simple behaviour and also flexible. Not sure how we deal with breaking changes for snekbox though.

ionite34 replied:
Yeah I agree this isn't ideal, I was just trying to make this work without breaking changes mostly.

The timeit situation mainly stems from the fact that it requires string input
```py
{
	"input": "[i for i in range(100)]",
	"args": ["-m", "timeit"]
}
->
python3 -m timeit "i for i in range(100)"
```
whereas something like `unittest` would requires a module/file
```py
python3 -m unittest main.py
```

So, not sure how we can really decide on snekbox side, maybe this should be something in the input request?

ionite34 replied:
Optionally we can just fix `input` to always write a file, and if the user wanted `-c` or `timeit` strings, require that to be in `args` instead?

ionite34 replied:
Implemented new api format in [ba6d0a8](https://github.com/python-discord/snekbox/pull/159/commits/ba6d0a8a10af687393134fc1e9662100ce67df52) which avoids parsing / guessing on snekbox's side in favor of the client providing all arguments and files. 

```json
"args": ["main.py"],
"files": [
    {
        "name": "main.py",
        "content": "print(1)"
    }
]
```

wookie184 reviewed (changes requested):
Not a proper review yet, just some comments from a quick look over.

wookie184 commented on snekbox/api/resources/eval.py:
```diff
@@ -61,7 +59,14 @@ def on_post(self, req: falcon.Request, resp: falcon.Response) -> None:
+        ...     "attachments": [
+        ...         {
+        ...             "name": "output.png",
+        ...             "mime": "image/png",
```
Same for size and compression type. If we're only using one compression type then there's no need to include it in the response (and I think supporting multiple compression types is overkill unless there's a real need for it). And with the size, is it really needed?

ionite34 replied:
Since we're planning on having a higher file size limit on the snekbox end and doing so on the bot end, it might be a waste of resources to uncompress a very large and well compressed content just to discover it's over our discord-side limit. Some 48MB of repetitive strings could compress to ~4KB, for example.

ionite34 replied:
Removed mime and compression in [ba6d0a8](https://github.com/python-discord/snekbox/pull/159/commits/ba6d0a8a10af687393134fc1e9662100ce67df52)

ionite34 commented on snekbox/memfs.py:
```diff
@@ -0,0 +1,151 @@
+    """Unmount and remove a tmpfs directory."""
+    tmp = NAMESPACE_DIR / name
+    subprocess.check_call(["umount", str(tmp)])
+    rmtree(tmp, ignore_errors=True)
```
It seems from my testing it still leaves the folder, should I do a `tmp.unlink()` instead then?

<img width="350" alt="image" src="https://user-images.githubusercontent.com/13956642/202793747-edcfc299-d642-42c3-a187-090aaf6f50f6.png">

wookie184 replied:
Ah, the directory left isn't the tmpfs, it's just the mountpoint directory created in `mount_tmpfs`. I think it would be better to use `tmp.rmdir()` and add a comment to make that clear.

ionite34 replied:
Resolved in [09bbc4f](https://github.com/python-discord/snekbox/pull/159/commits/09bbc4f11f32d91dfa20aeb3da77a5e2a054f49d)

wookie184 commented on snekbox/memfs.py:
```diff
@@ -0,0 +1,154 @@
+    """Create and mount a tmpfs directory."""
+    tmp = NAMESPACE_DIR / name
+    tmp.mkdir()
+    tmp.chmod(0o711)
```
Why is a chmod needed here? (also all the other chmods)

ionite34 replied:
I think this was a remanent when there used to be exposed access to this folder. Removed in a0c908a. The chmods for the home directory would have to stay because that needs to be read/write.

MarkKoz commented on snekbox/memfs.py:
```diff
@@ -0,0 +1,137 @@
+        unmount_tmpfs(self.name)
+
+        if not self.path.exists():
+            with self.assignment_lock:
```
Why is a lock necessary?

ionite34 replied:
I was thinking there might be issues if we had multiple threads? Though I suppose we don't have that currently. I don't think it hurts to make it thread safe for the time being though?

MarkKoz replied:
Gunicorn is configured with multiple workers, though right now I cannot recall if those are threads or processes. My vote is for simplicity, so a lock shouldn't be included if it's not necessary. Can you investigate further? If it is needed, then there should also be a test for whatever bad situation this lock prevents.

ionite34 replied:
It seems the `path.mkdir()` in `mount_tmpfs` should already be sufficient as the python unix implementation appears to be free from race conditions.

Removed set and semaphore in [8c897dc](https://github.com/python-discord/snekbox/pull/159/commits/8c897dc0e86f4c6cb5e09e376c6aa40d4093ed7d) in favor of catching the `FileExistsError` instead.

Added tests in [b74b813](https://github.com/python-discord/snekbox/pull/159/commits/b74b813ae6ac2ebf374cc12bb0e9689f49776f2e) to ensure this new approach works fine in multithread conditions.

MarkKoz commented on snekbox/nsjail.py:
```diff
@@ -163,37 +193,49 @@ def python3(
-                *py_args,
-                code,
-            )
+                # Filter out empty strings (causes issues with python cli)
```
Only empty strings at the start should be removed. Empty strings elsewhere could conceivably be valid arguments for the script.

MarkKoz commented on snekbox/nsjail.py:
```diff
@@ -163,37 +193,49 @@ def python3(
+            # Write files if any
+            for file in files:
+                file.save_to(fs.home)
+                log.info(f"Created file at [{(fs.home / file.name)!r}].")
```
The brackets seem unnecessary; the path is already going to be quoted.
```suggestion
                log.info(f"Created file at {(fs.home / file.name)!r}.")
```

MarkKoz commented on snekbox/nsjail.py:
```diff
@@ -163,37 +193,49 @@ def python3(
             msg = "Executing code..."
             if DEBUG:
-                msg = f"{msg[:-3]}:\n{textwrap.indent(code, '    ')}\nWith the arguments {args}."
+                msg = f"{msg[:-3]}: With the arguments {args}."
```
```suggestion
                msg = f"{msg[:-3]} with the arguments {args}."
```

MarkKoz commented on snekbox/nsjail.py:
```diff
@@ -129,16 +137,24 @@ def _consume_stdout(self, nsjail: subprocess.Popen) -> str:
-        which is the last argument. By default, it's "-c", which executes the code given.
+        which is the last argument.
+
+        If `use_file` is True, `code` will be written to a file the last argument will be
+        the file name `main.py`. If False, `code` will be passed as the last argument.
+        The default of None will use a file unless `c` is in `py_args`.
```
This seems out of date.

ionite34 replied:
Updated in [f12f8f3](https://github.com/python-discord/snekbox/pull/159/commits/f12f8f3915620cf5f20ef5c36e488517a4e73e4b)

MarkKoz commented on snekbox/snekio.py:
```diff
@@ -0,0 +1,108 @@
+        if any(set(part) == {"."} for part in parts):
+            raise IllegalPathError(f"File path '{name}' may not use traversal ('..')")
+
+        match data.get("content-encoding"):
+            case "base64":
+                content = b64decode(data["content"])
+            case None | "utf-8" | "":
+                content = data["content"].encode("utf-8")
+            case _:
+                raise ParsingError(f"Unknown content encoding '{data['content-encoding']}'")
```
Let's not overcomplicate this. Similar to how the response always uses b64 for files, the request should also always use b64 for files.

ionite34 replied:
Made request accept base64 only, removed `content-encoding` parameter, in [4787a94](https://github.com/python-discord/snekbox/pull/159/commits/4787a9497be4f460e028031ddad4cd9d8388b84f)

MarkKoz commented on snekbox/memfs.py:
```diff
@@ -0,0 +1,137 @@
+        """Return a list of attachments in the tempdir."""
+        count = 0
+        # Look for any file starting with `output`
+        for file in self.home.glob("output*"):
```
Thoughts on requiring writing to an "output" directory rather than requiring file names to be prefixed with `output`?

ionite34 replied:
Hm, I think that sounds reasonable. I was originally trying to avoid creating any other directories in the instance /home since now there's a potential FileExistsError if an user provides a file called `output` in the request. Though I suppose we could check for that as a validation?

ionite34 replied:
Implemented since [78b4b6a](https://github.com/python-discord/snekbox/pull/159/commits/78b4b6af18a40db3162aad56eb726a26c5a74e8c)

MarkKoz commented on snekbox/nsjail.py:
```diff
@@ -129,16 +137,24 @@ def _consume_stdout(self, nsjail: subprocess.Popen) -> str:
+        self,
+        py_args: Iterable[str],
+        files: Iterable[FileAttachment] = (),
+        *,
```
The placement of this `*` seems arbitrary. It would make sense for it to be before `py_args`, but I would be fine if it was just removed entirely at this point.

ionite34 replied:
Removed `*` in [4787a94](https://github.com/python-discord/snekbox/pull/159/commits/4787a9497be4f460e028031ddad4cd9d8388b84f)

MarkKoz commented on snekbox/nsjail.py:
```diff
@@ -152,8 +168,22 @@ def python3(
+                # Mount `home` with Read/Write access
+                "--bindmount",
+                f"{fs.home}:home",
+                # Set cwd to temp dir
+                "--cwd",
+                "home",
+                # Some packages rely on the HOME env variable
+                "--env",
+                "HOME=home",
```
Why not set these in the config file instead?

ionite34 replied:
Moved cwd and home env to config file in [c30c0eb](https://github.com/python-discord/snekbox/pull/159/commits/c30c0eb443f3ab759e9dfcbfee5bac6f4c428048)

MarkKoz commented on snekbox/__main__.py:
```diff
@@ -37,7 +37,7 @@ def parse_args() -> argparse.Namespace:
     """Evaluate Python code through NsJail."""
     args = parse_args()
-    result = NsJail().python3(args.code, nsjail_args=args.nsjail_args, py_args=args.py_args)
+    result = NsJail().python3(py_args=[*args.py_args, args.code], nsjail_args=args.nsjail_args)
```
`py_args` needs to default to `-c` here to preserve the original behaviour, since `python3()` no longer has it as a default. (I think this is the most convenient default when using it via the CLI for testing).

ionite34 replied:
I think the argparse default here already makes `-c` the default? The previous unit tests for main also work fine.

MarkKoz commented on snekbox/snekio.py:
```diff
@@ -0,0 +1,108 @@
+        path = Path(name)
+        parts = path.parts
+
+        if path.is_absolute() or set(parts[0]) & {"\\", "/"}:
+            raise IllegalPathError(f"File path '{name}' must be relative")
```
`is_absolute` covers the `/` case, and `"\\"` is of no concern since it's not considered absolute on Posix.

ionite34 replied:
Only using `is_absolute` since [00b7e3b](https://github.com/python-discord/snekbox/pull/159/commits/00b7e3bd13b7914a44c5af00ff6f8da9b48fa9c5)

MarkKoz commented on snekbox/snekio.py:
```diff
@@ -0,0 +1,108 @@
+        if path.is_absolute() or set(parts[0]) & {"\\", "/"}:
+            raise IllegalPathError(f"File path '{name}' must be relative")
+
+        if any(set(part) == {"."} for part in parts):
+            raise IllegalPathError(f"File path '{name}' may not use traversal ('..')")
```
This only checks a single dot. Shouldn't it also check `..`? I wonder if there is a more convenient way to detect this. Maybe https://stackoverflow.com/a/45190125 ?

ionite34 replied:
Implemented the pathlib solution since [00b7e3b](https://github.com/python-discord/snekbox/pull/159/commits/00b7e3bd13b7914a44c5af00ff6f8da9b48fa9c5) as detailed in the stackoverflow post. Should be better since now we only disallow traversal beyond root.

ionite34 commented on tests/test_nsjail.py:
```diff
@@ -272,14 +336,14 @@ def test_large_output_is_truncated(self):
 
     def test_nsjail_args(self):
-        args = ("foo", "bar")
+        args = ["foo", "bar"]
```
It's due to the fact I've changed the call within the function to a list and not a tuple since I had to do a couple extend/appends. Wasn't sure whether to cast back to a tuple or just leave it and ended up just leaving it. Do you think it's better if we cast it to a tuple at the end?

MarkKoz replied:
Thanks for clarifying. No objections to keeping it a list.

ionite34 commented on docker-compose.yml:
```diff
@@ -8,7 +8,7 @@ services:
     pull_policy: never
     ports:
-     - 8060:8060
+     - "8060:8060"
```
Docker recommends to only use strings for port numbers https://docs.docker.com/compose/compose-file/compose-file-v3/#short-syntax-1

MarkKoz replied:
Thanks. YAML is very cool and good.

ionite34 commented on snekbox/api/resources/eval.py:
```diff
@@ -38,8 +49,6 @@ def on_post(self, req: falcon.Request, resp: falcon.Response) -> None:
         Evaluate Python code and return stdout, stderr, and the return code.
 
         A list of arguments for the Python subprocess can be specified as `args`.
-        Otherwise, the default argument "-c" is used to execute the input code.
-        The input code is always passed as the last argument to Python.
```
Seems reasonable yeah, I'll look into compatible support for "input"

ionite34 replied:
should we have args implicitly include an "-c" default if "input" is non-empty?

MarkKoz replied:
If `input` exists, it should become the last Python argument. If `input` exists and `args` does not exist, then `args` should default to `["-c"]` (in addition to the `input` becoming the last Python argument).

ionite34 replied:
Honestly after attempting an implementation, this seems to just increase complexity and error-prone code (like the fact that our specification will be an 'either-or-but-at-least-one' requirement of `args` and `input`. The initial inclusion of `input` was already largely redundant as it is directly appended to `args`, so `args` can achieve all of the previous usages of `input` but without the ambiguous parsing and dynamic default of "-c" that we do on our end. 

In the future, if we support requests specifying user input via `sys.argv`, the name occupied by `input` will be more ambiguous and confusing. 

In general the deprecation of `input` at this stage seems like a reasonable path. If we truly need to support users beyond python bot, imo the usual case is maintaining a versioned endpoint that retains previous behavior and specifications fully, instead of retaining it as an optional parameter in the same endpoint (Which I've never really seen as a path taken by any major public API).

For the CLI interface, the simplicity in typing the code directly could be made, but for an API, we're retaining a parameter (input) just to be later conditionally appended to an already existing parameter (args) which has a default that conditionally exists based on whether the new parameter is empty or not. With all of this being able to be avoided from the requester supplying a single explicit "-c" in `args`.

Considering the additional keys in the response is already potentially breaking, and our numerous breaking library-side changes (such as the return types and arguments of `python3`), I don't really see a library side breaking concern more than any other change. We should increment a major version, and downstream users can upgrade when they're ready.

ionite34 replied:
I've added compatible support for input in the API route in 84e73c1, so it is backwards compatible now. Added some associated unit tests to ensure the old `input` style still works.

Currently the API route requires either `args` or `input`. If `input` exists, `args` will have a default of `-c` if it is otherwise empty.

ionite34 commented on tests/test_nsjail.py:
```diff
@@ -99,7 +103,7 @@ def f():
 
     def test_read_only_file_system(self):
-        for path in ("/", "/etc", "/lib", "/lib64", "/snekbox", "/usr"):
+        for path in ("/etc", "/lib", "/lib64", "/snekbox", "/usr"):
```
Seems accidental. Re-added in [291119d](https://github.com/python-discord/snekbox/pull/159/commits/291119dc45817a01f29afcde0328a53f9102df3c)

ionite34 commented on tests/test_nsjail.py:
```diff
@@ -113,6 +117,64 @@ def test_read_only_file_system(self):
+        self.assertIn("No space left on device", result.stdout)
+        self.assertEqual(result.stderr, None)
+        # Restore the original size
+        self.nsjail.memfs_instance_size = backup
```
Make instance size lower global for the tests in [a1c614d](https://github.com/python-discord/snekbox/pull/159/commits/a1c614d7e950a6329bfaeec7ef1467553bd96a37)

ionite34 commented on snekbox/snekio.py:
```diff
@@ -0,0 +1,89 @@
+            raise AttachmentError(
+                f"File {file.name} too large: {sizeof_fmt(size)} "
+                f"exceeds the limit of {sizeof_fmt(max_size)}"
+            )
```
Removed file size limit and associated errors, added configurable file parse time limit in [73a5501](https://github.com/python-discord/snekbox/pull/159/commits/73a5501e04706e190ee7d97ac006924b0e90463a)

ionite34 commented on snekbox/snekio.py:
```diff
@@ -0,0 +1,61 @@
+
+    def to_dict(self) -> dict[str, str]:
+        """Convert the attachment to a dict."""
+        cmp = zlib.compress(self.content)
```
I've removed the zlib compression completely and content-encoding parsing in [4787a94](https://github.com/python-discord/snekbox/pull/159/commits/4787a9497be4f460e028031ddad4cd9d8388b84f)

I think since compression doesn't seem essential at this point, for the file sizes we're dealing with, (and won't change API if we implement server-layer compression), I'll probably leave that for a future PR.

ionite34 commented on config/snekbox.cfg:
```diff
@@ -104,11 +104,13 @@ mount {
     rw: false
 }
 
+rlimit_fsize: 134217728
```
Set to -1 in [80d4c45](https://github.com/python-discord/snekbox/pull/159/commits/80d4c45ae6bd152dd7c1dfcfbaa1995dd0018954)

MarkKoz replied:
Ah, I did not realise NsJail had a default. Unlimited is fine, I think... Ultimately we're concerned without overall disk usage being limited. If that's somehow bypassed or misconfigured, I don't see how a malicious user creating 1 large file (when rlimit is -1) vs. many small files (when rlimit is set) would have a significant impact.

ionite34 commented on snekbox/api/resources/eval.py:
```diff
@@ -53,33 +62,57 @@ def on_post(self, req: falcon.Request, resp: falcon.Response) -> None:
+        ...     "returncode": 0,
+        ...     "attachments": [
+        ...         {
+        ...             "name": "output.png",
```
Renamed to path in 4787a94

ionite34 commented on snekbox/api/resources/eval.py:
```diff
@@ -53,33 +62,57 @@ def on_post(self, req: falcon.Request, resp: falcon.Response) -> None:
+        ...    "args": ["main.py"],
+        ...    "files": [
+        ...        {
+        ...            "name": "main.py",
```
Renamed to path in 4787a94

ionite34 commented on snekbox/api/resources/eval.py:
```diff
@@ -53,33 +62,57 @@ def on_post(self, req: falcon.Request, resp: falcon.Response) -> None:
+        ...             "size": 57344,
+        ...             "compression": "zlib",
+        ...             "content": "eJzzSM3...="  # Base64-encoded
+        ...         }
```
Updated example along with changes in 4787a94

MarkKoz commented on tests/test_nsjail.py:
```diff
@@ -63,9 +85,12 @@ def test_subprocess_resource_unavailable(self):
         self.assertIn("Resource temporarily unavailable", result.stdout)
+        # Also expect n-1 processes to be opened
+        res = result.stdout.split("T")[0].strip().split("\n")  # ['2', '3', '4'] for n=4
+        self.assertEqual(res, [*map(str, range(2, max_pids + 1))])
```
Oh, this is checking the PIDs being printed! I didn't see that those get printed at first; I got lost trying to understand the string parsing happening here. A comment explaining that PIDs are being extracted from stdout would help.

ionite34 replied:
Refactored and added comments to make this clearer in [2b20b85](https://github.com/python-discord/snekbox/pull/159/commits/2b20b854b027c1ba27aa85b1dcef96c4bfa7b6b9)

MarkKoz commented on snekbox/nsjail.py:
```diff
@@ -205,8 +205,9 @@ def python3(
-                *(arg for arg in py_args if arg),
+                # Filter out empty strings at start of iterable
+                # (causes issues with python cli)
+                *(arg for i, arg in enumerate(py_args) if (arg or i > 0)),
```
This doesn't cover the case of multiple empty strings at the start. I also think writing a test for this would be beneficial.

ionite34 replied:
Created function `iter_lstrip` for this task, and additional unit tests since [621c835](https://github.com/python-discord/snekbox/pull/159/commits/621c835f28cd3fcd1e5c3cccae9b5057647f9fa1)

MarkKoz commented on snekbox/nsjail.py:
```diff
@@ -227,13 +239,14 @@ def python3(
-                log.info(f"Failed to parse attachments: {err}")
+            except TimeoutError as err:
+                log.info(f"Exceeded time limit in parsing attachments: {err}")
                 return EvalResult(args, returncode, f"AttachmentError: {err}")
```
The `AttachmentError: ` prefix seems misleading because no such exception type still exists. I don't think it's necessary.

ionite34 replied:
Changed in [9400107](https://github.com/python-discord/snekbox/pull/159/commits/9400107e00d37b1809dd1ff8276a6b5bbba9dfdd)

MarkKoz commented on snekbox/nsjail.py:
```diff
@@ -227,13 +239,14 @@ def python3(
                 log.info(f"Found {len(attachments)} attachments.")
-            except AttachmentError as err:
-                log.info(f"Failed to parse attachments: {err}")
+            except TimeoutError as err:
+                log.info(f"Exceeded time limit in parsing attachments: {err}")
```
I have mixed feelings towards having a timeout. I'm leaning towards keeping it simple and not having a timeout. It should be possible to prevent long parsing times by just limiting the amount of files, for which we already have a setting.

Furthermore, it seems unintuitive that the process may have an OK exit code with some stdout available, but it's getting overwritten by an error, and the HTTP response is still 200. This is also a concern I had with the previous code that was erroring out when a file size limit was exceeded.

What are everyone's thoughts on having a timeout for file parsing? Is it beneficial given the other measures we have in place?

ionite34 replied:
My main concern is, if we allow nested directories in outputs, someone could create a very deeply nested directory as a denial of service attack. 

In general this is a potentially lengthy part of server side code execution that can be entirely controlled by the user.

ionite34 replied:
I agree that the response feels unintuitive though. Maybe it can just be a 500 error? This is more of a catastrophic fail-safe than something that is going to come up often.

ionite34 replied:
I have changed the returncode of the TimeoutError to `None` to be in-line with the previous snekbox stdout errors.

MarkKoz commented on snekbox/nsjail.py:
```diff
@@ -38,16 +48,18 @@ def __init__(
-        max_attachment_size: int | None = None,
+        files_limit: int = 100,
+        files_timeout: float = 15,
+        files_pattern: str = "output*",
```
I think it's a good idea that this is configurable now.

ionite34 commented on snekbox/memfs.py:
```diff
@@ -0,0 +1,137 @@
+
+        size: Size limit of each tmpfs instance in bytes
+        """
+        self.path: Path | None = None
```
Changed in [8c897dc](https://github.com/python-discord/snekbox/pull/159/commits/8c897dc0e86f4c6cb5e09e376c6aa40d4093ed7d) to raise a `RuntimeError` on most attribute access when not initialized through `__enter__`

ionite34 commented on snekbox/memfs.py:
```diff
@@ -0,0 +1,137 @@
+            "tmpfs",
+            str(tmp),
+        ]
+    )
```
Refactored to use syscall for mount and unmount in [0387793](https://github.com/python-discord/snekbox/pull/159/commits/0387793222cb1c0feb5644ba8ba8c26caa84c7b2)

ionite34 commented on snekbox/snekio.py:
```diff
@@ -0,0 +1,89 @@
+        content = b64encode(cmp).decode("ascii")
+        return {
+            "name": self.name,
+            "size": self.size,
```
For our usage this saves python bot having to decode a base64 content string if size already exceeds our limit, which is not a huge optimization but could be a useful information which is free for us to include.

MarkKoz commented on tests/test_libmount.py:
```diff
@@ -0,0 +1,135 @@
+        ]
+        for flag in flags:
+            with self.subTest(flag=flag), self.get_mount() as path:
+                libmount.unmount(path, flag)
```
What is the assertion here?

MarkKoz commented on tests/test_nsjail.py:
```diff
@@ -135,25 +223,27 @@ def test_sigsegv_returns_139(self):  # In honour of Juan.
-        result = self.nsjail.python3("\0")
+        # This error does not occur without -c, where it
+        # would be a normal SyntaxError.
+        result = self.nsjail.python3(["-c", "\0"])
```
This can use the `eval_code` utility, right?

ionite34 replied:
Yup, using `eval_code` with [fca97ce](https://github.com/python-discord/snekbox/pull/159/commits/fca97cecea5040dac8ab02046cffac0713f78dd2)

MarkKoz commented on snekbox/memfs.py:
```diff
@@ -0,0 +1,148 @@
+    @cached_property
+    def path(self) -> Path:
+        """Returns the path of the MemFS."""
+        if self._path is None:
+            raise RuntimeError("MemFS accessed before __enter__.")
```
Sorry to sort of "go back" on this, but it dawned on me: why is this being set in `__enter__` instead of `__init__`? Most libraries I've seen support using an object without a context manager (requiring a manual call to some public cleanup function). Wouldn't it make more sense for all of `__enter__` to just be in `__init__` instead? This current situation of being able to instantiate an object in an invalid state just feels really wrong. On the other hand, having side effects when instantiation an object also doesn't sound great, but the former seems like the worse of the two.

I don't want to block the PR on this. This is all internal stuff we can revisit at a later time if you prefer.

ionite34 replied:
Yeah I agree, I think the context manager being optional is more in-line with conventions. I've changed it to be initialized during `__init__`, with a `weakref.finalize` running the cleanup on garbage collection with a `ResourceWarning` if the context manager is not used, modeled after https://github.com/python/cpython/blob/main/Lib/tempfile.py#L844

MarkKoz commented on snekbox/nsjail.py:
```diff
@@ -129,16 +165,18 @@ def _consume_stdout(self, nsjail: subprocess.Popen) -> str:
-        `py_args` are arguments to pass to the Python subprocess before the code,
-        which is the last argument. By default, it's "-c", which executes the code given.
+        Args:
+            py_args: Arguments passed to python3.
+            files: FileAttachments to be written to the sandbox prior to run.
+            nsjail_args: Overrides options in the NsJail config, added to NsJail call.
```
```suggestion
            py_args: Arguments to pass to Python.
            files: FileAttachments to write to the sandbox prior to running Python.
            nsjail_args: Overrides for the NsJail configuration.
```

MarkKoz commented on snekbox/snekio.py:
```diff
@@ -0,0 +1,101 @@
+        return self.content.encode("utf-8")
+
+    def save_to(self, directory: Path | str) -> None:
+        """Save the attachment to a path directory."""
```
```suggestion
        """Write the attachment to a file in `directory`."""
```

MarkKoz replied:
The `directory` parameter should be included in the docstring.

MarkKoz commented on snekbox/nsjail.py:
```diff
@@ -2,26 +2,41 @@
 
 
+def iter_lstrip(iterable: Iterable[_T]) -> Generator[_T, None, None]:
+    """Removes leading falsy objects from an iterable."""
```
```suggestion
    """Remove leading falsy objects from an iterable."""
```

MarkKoz reviewed (changes requested):
I'm concerned that I'm leaving too many comments on non-blockers. Let me know if you'd prefer to defer some comments to a future PR (which doesn't necessarily have to be implemented by you). I don't want to create a draining review cycle for you.

MarkKoz commented on snekbox/nsjail.py:
```diff
@@ -19,18 +20,35 @@
+
+    Returns:
+        List of FileAttachments sorted lexically by path name.
+    """
```
Okay this comment is a bit outdated now that I noticed you changed the docstring style. Still, I think this could just be changed to the one liner I suggested above.

ionite34 replied:
Refactored `parse_files` as an instance function `files_list` of `MemFS` in [8272f88](https://github.com/python-discord/snekbox/pull/159/commits/8272f88c3b50cc3912ae087b4f7eec506a5abaac). Made argument naming between `files` and `files_list` more coherent.

ionite34 commented on tests/test_nsjail.py:
```diff
@@ -361,11 +361,18 @@ def test_nsjail_args(self):
+            ["", "-m", "timeit"],
+            ["", "", "-m", "timeit"],
+            ["", "", "", "-m", "timeit"],
+        ]
```
Included since [2da0ddf](https://github.com/python-discord/snekbox/pull/159/commits/2da0ddf14625cfd8037c198881ab789827746c40)

ionite34 commented on tests/test_libmount.py:
```diff
@@ -0,0 +1,135 @@
+class LibMountTests(TestCase):
+    def setUp(self):
+        self.temp_dir = Path("/tmp/snekbox-test")
+        self.temp_dir.mkdir(exist_ok=True, parents=True)
```
Changed to use `setUpClass` and `tearDownClass` and cleanups in [f4494c3](https://github.com/python-discord/snekbox/pull/159/commits/f4494c307eb5cf4200fd678dff90f1c2b5a5775e)

ionite34 commented on snekbox/snekio.py:
```diff
@@ -0,0 +1,101 @@
+
+
+@dataclass
+class FileAttachment(Generic[T]):
```
Refactored as non-generic in [407e6b2](https://github.com/python-discord/snekbox/pull/159/commits/407e6b2079ea9f73f52e4972147620d765d96349)

ionite34 commented on snekbox/snekio.py:
```diff
@@ -0,0 +1,101 @@
+        """Return the attachment as bytes."""
+        if isinstance(self.content, bytes):
+            return self.content
+        return self.content.encode("utf-8")
```
Refactored as non-generic in [407e6b2](https://github.com/python-discord/snekbox/pull/159/commits/407e6b2079ea9f73f52e4972147620d765d96349)

ionite34 commented on snekbox/snekio.py:
```diff
@@ -0,0 +1,101 @@
+            file.write_bytes(self.content)
+
+    @cached_property
+    def json(self) -> dict[str, str]:
```
I've renamed to `as_dict` since I feel the fact that this is a property and not a method, it might be confusing to share the same name `to_dict` as the other method ones. [fabaae9](https://github.com/python-discord/snekbox/pull/159/commits/fabaae9472321ec35a0736b96ea4af9b72d30eae)

ionite34 commented on snekbox/snekio.py:
```diff
@@ -0,0 +1,101 @@
+    @classmethod
+    def from_dict(cls, data: dict[str, str]) -> FileAttachment[bytes]:
+        """Convert a dict to an attachment."""
+        path = safe_path(data["path"])
```
I think the only invalid path is a single `/` on linux? And that would be caught by as an absolute path. File writing is also in a `try` block with an associated `OSError` report.

Though I have added handling for absolute paths (any path beginning with `/` or `\0`) and unit tests for these in [7e7c0d6](https://github.com/python-discord/snekbox/pull/159/commits/7e7c0d6f5465f965b3ee31fc4f723ec23cc5ad94)

I am not sure if we can do traversal checks with regex. And I think we should still keep the current pathlib checks for absolute and traversal (since the json schema validator is not used for `__main__` calls and library use)

ionite34 replied:
Added additional unit tests in [dffd9c8](https://github.com/python-discord/snekbox/pull/159/commits/dffd9c8100230d6e8df8c0261fbd5ae8dbd0ceb2)

ionite34 replied:
Updated regex to
```regex
^(?!/)(?!.*\\0).*$
```
This should disallow anything beginning with `/` or `\0` contained anywhere.

ionite34 commented on snekbox/libmount.py:
```diff
@@ -0,0 +1,81 @@
+        raise OSError(errno, f"Error mounting {target}: {os.strerror(errno)}")
+
+
+def unmount(target: Path | str, flags: UnmountFlags | int = UnmountFlags.MNT_DETACH) -> None:
```
I think that's better yeah, renamed to `filesystem` in [e72032f](https://github.com/python-discord/snekbox/pull/159/commits/e72032fc3090d6e5d5d961a3aafaeec4cce787d7)

ionite34 commented on snekbox/snekio.py:
```diff
@@ -0,0 +1,101 @@
+    def from_dict(cls, data: dict[str, str]) -> FileAttachment[bytes]:
+        """Convert a dict to an attachment."""
+        path = safe_path(data["path"])
+        content = b64decode(data.get("content", ""))
```
Added error handling in [0442a7c](https://github.com/python-discord/snekbox/pull/159/commits/0442a7cec24bff9985fb8cf59b0fb2604544eaa8), additional base64 unit tests in [dffd9c8](https://github.com/python-discord/snekbox/pull/159/commits/dffd9c8100230d6e8df8c0261fbd5ae8dbd0ceb2)

ionite34 commented on snekbox/api/resources/eval.py:
```diff
@@ -53,33 +65,57 @@ def on_post(self, req: falcon.Request, resp: falcon.Response) -> None:
+        except IllegalPathError as e:
+            raise falcon.HTTPBadRequest(
+                title="Request file path failed validation", description=str(e)
+            )
```
I've added schema level validation for `\0` and paths starting with `/`, though I think we'll still need this for the traversal checks and for `__main__` usage.

ionite34 commented on snekbox/nsjail.py:
```diff
@@ -35,11 +50,32 @@ def __init__(
+            read_chunk_size: Size of the read buffer in bytes.
+            memfs_instance_size: Size of the tmpfs instance in bytes.
+            files_limit: Maximum number of files to parse for attach.
+            files_timeout: Maximum time in seconds to wait for files to be written / read.
```
Should be yeah, clarified in [2ba9431](https://github.com/python-discord/snekbox/pull/159/commits/2ba94312e60300b136b55781c8f8e3149e8bb0ad)

ionite34 commented on snekbox/nsjail.py:
```diff
@@ -182,15 +182,12 @@ def python3(
             nsjail_args = (
+                # Set fslimit to unlimited, cannot be set in cfg
+                # due to upstream protobuf parsing issue
+                "--rlimit_fsize=inf",
```
Thanks for that investigation. `rlimit_fsize_type: INF` works perfect.

[15e8df0](https://github.com/python-discord/snekbox/pull/159/commits/15e8df09356d53639dbe50034b82b1356dda6af8)

MarkKoz commented on tests/test_memfs.py:
```diff
@@ -0,0 +1,67 @@
+        super().setUp()
+        self.logger = logging.getLogger("snekbox.memfs")
+        self.logger.setLevel(logging.WARNING)
+        warnings.filterwarnings(
+            "ignore",
+            ".*Implicitly cleaning up.*",
+            ResourceWarning,
+            "snekbox.memfs",
+        )
```
I think this filter may persist for other test cases. Is that a concern?

ionite34 replied:
This was due to the multithread test relying on implicit cleanup. I've since changed it to use contextlib.ExitStack instead to call `__exit__`, so this is not necessary and removed in [d33a55d](https://github.com/python-discord/snekbox/pull/159/commits/d33a55d775c9f15b6cd9089c7e2de3e827f0b837)

MarkKoz commented on snekbox/memfs.py:
```diff
@@ -0,0 +1,151 @@
+
+
+class MemFS:
+    """A memory temporary file system."""
```
```suggestion
    """An in-memory temporary file system."""
```

MarkKoz commented on snekbox/memfs.py:
```diff
@@ -0,0 +1,151 @@
+
+    def __init__(self, instance_size: int, root_dir: str | Path = "/memfs") -> None:
+        """
+        Initialize a memory temporary file system.
```
```suggestion
        Initialize an in-memory temporary file system.
```

MarkKoz commented on tests/api/test_eval.py:
```diff
@@ -5,12 +5,16 @@ class TestEvalResource(SnekAPITestCase):
-        self.assertEqual("output", result.json["stdout"])
-        self.assertEqual(0, result.json["returncode"])
+        cases = [
+            {"args": ["-c", "print('output')"]},
+            {"input": "print('hello')"},
```
We should also test a case where both input and args are provided.

ionite34 replied:
Added in [95e3e18](https://github.com/python-discord/snekbox/pull/159/commits/95e3e18ac93709257840f8e5c36e6543678931f5)

MarkKoz commented:
Does the `output` directory have to be created beforehand or is that already being done somewhere?

MarkKoz commented on snekbox/nsjail.py:
```diff
@@ -35,11 +51,32 @@ def __init__(
+            memfs_instance_size: Size of the tmpfs instance in bytes.
+            files_limit: Maximum number of output files to parse.
+            files_timeout: Maximum time in seconds to wait for output files to be read.
+            files_pattern: Pattern to match files to attach.
```
Would it be possible to make the output directory part of the files_pattern? Would there be a problem with ambiguity in whether the directory should be created by NsJail or by the user?

ionite34 replied:
Well currently we have the glob start at the output folder, otherwise we'd have the output folder in the glob result and have to strip it for sending.

If we allow the output to be in the glob, the folder will still be created by NSJail so an end user will still need to change the code to have a different output folder name for example.

MarkKoz replied:
I don't understand your answer, but I think it's because my question was unclear.

Currently, the output directory is hard-coded to `output`. I'm wondering if it'd be possible to make that configurable by allowing a user to specify the directory name in this `files_pattern` glob. However, I think this could result in ambiguity regarding what should be the path snekbox creates, and what should the end-user have to manually create. 

In any case, we can revisit it later. Not a blocker for the PR since I don't believe making this configurable in the future will be a breaking change.

ionite34 replied:
Ah I see what you mean. I think implicitly creating paths in the glob might be a bit complex (it also would allow some arbitrary traversal and other situations) I've made the name of `home` and `output` configurable now in [c3a43ac](https://github.com/python-discord/snekbox/pull/159/commits/c3a43ac3b254efc999727c9fd15aac64d3203162)

So to not use output, `memfs_output=""` can be set to set the output glob root to home, or any other path.

ionite34 commented on README.md:
```diff
@@ -60,10 +63,42 @@ The main features of the default configuration are:
 
 NsJail is configured through [`snekbox.cfg`]. It contains the exact values for the items listed above. The configuration format is defined by a [protobuf file][7] which can be referred to for documentation. The command-line options of NsJail can also serve as documentation since they closely follow the config file format.
 
+### Memory File System
```
Updated in [6ec4dd2](https://github.com/python-discord/snekbox/pull/159/commits/6ec4dd2d3026d1fe8e567b7eaf723d82811cdcf1)

ionite34 commented on snekbox/filesystem.py:
```diff
@@ -0,0 +1,88 @@
+    KB = 1024
+    MB = 1024**2
+    GB = 1024**3
+    TB = 1024**4
```
Renamed in [4abc988](https://github.com/python-discord/snekbox/pull/159/commits/4abc988c6e5b9ede6cffe372d3252d9863c8eef0)

ionite34 commented:
> Does the `output` directory have to be created beforehand or is that already being done somewhere?

It's done here 
https://github.com/python-discord/snekbox/blob/bytes-output/snekbox/memfs.py#L54

ionite34 commented on README.md:
```diff
@@ -60,10 +63,42 @@ The main features of the default configuration are:
+On each execution, the host will mount an instance-specific `tmpfs` directory,
+this is used as a limited read-write folder for the sandboxed code. There is no
+access of any kind to other files or directories on the host system or container,
+as only this new directory is mounted to NSJail.
```
Yeah I think I need to rephrase this. Updated in [6ec4dd2](https://github.com/python-discord/snekbox/pull/159/commits/6ec4dd2d3026d1fe8e567b7eaf723d82811cdcf1)

ionite34 commented on README.md:
```diff
@@ -60,10 +63,42 @@ The main features of the default configuration are:
+Files written to the `output` subfolder will be parsed and returned as
+a list of `FileAttachment` objects in `EvalResult.files` from the `python3` function.
+
+To send files to snekbox, it can be included as the `files` parameter of `python3`.
```
Removed in [8c7da5a](https://github.com/python-discord/snekbox/pull/159/commits/8c7da5a32a6959409f917fb32ba331fe5132751e)

ionite34 commented on README.md:
```diff
@@ -60,10 +63,42 @@ The main features of the default configuration are:
+
+Within the `/eval` route, files are attached or returned under the `files` key.
+
+> See [api/resources/eval.py](snekbox/api/resources/eval.py) for additional API schema information.
```
Yup, removed in [5c588f7](https://github.com/python-discord/snekbox/pull/159/commits/5c588f7832eebfffc5886f25f48f10514b51a4fe)

ionite34 commented on snekbox/memfs.py:
```diff
@@ -114,28 +89,46 @@ def mkdir(self, path: Path | str, chmod: int = 0o777) -> Path:
+        preload_dict: bool = False,
+    ) -> list[FileAttachment]:
+        """
+        Returns a sorted list of output files found in the MemFS.
```
Updated in [5d179ba](https://github.com/python-discord/snekbox/pull/159/commits/5d179ba5ee7e2d0c8808b8a7b90f6a560ab2acf0)

ionite34 commented on tests/test_libmount.py:
```diff
@@ -1,17 +1,23 @@
+
+    @classmethod
+    def tearDownClass(cls):
+        rmtree(cls.temp_dir, ignore_errors=True)
```
Actually speaking of `tempfile`, that's probably a better case, I've refactored this to use `tempfile.TemporaryDirectory` instead of manually doing this [62954cd](https://github.com/python-discord/snekbox/pull/159/commits/62954cd078cbcb4719cb1a1b726fe3672a17c495)

MarkKoz commented on snekbox/api/resources/eval.py:
```diff
@@ -36,8 +36,8 @@ class EvalResource:
                             "type": "string",
-                            # Disallow single absolute paths, and null bytes
-                            "pattern": r"^[^/\\0].*",
+                            # Disallow starting with / or containing \0 anywhere
+                            "pattern": r"^(?!/)(?!.*\\0).*$",
```
Absolute case seems to be covered in tests but we do need to test the null byte as well.

ionite34 replied:
Added additional tests in [55a9223](https://github.com/python-discord/snekbox/pull/159/commits/55a92231d83439b31441945799f713c25b569278) and checks to make sure separate `\` and `0` are allowed.

MarkKoz commented on README.md:
```diff
@@ -60,10 +63,28 @@ The main features of the default configuration are:
 
+### Memory File System
+
+On each execution, the host will mount an instance-specific `tmpfs` drive, this is used as a limited read-write folder for the sandboxed code. There is no access to other files or directories on the host container beyond the other read-only mounted system folders. Instance file systems are isolated, it is not possible for sandboxed code to access another instance's writeable directory.
```
This is a comma splice. Use a semicolon or a full stop to create a separate sentence.
```suggestion
On each execution, the host will mount an instance-specific `tmpfs` drive, this is used as a limited read-write folder for the sandboxed code. There is no access to other files or directories on the host container beyond the other read-only mounted system folders. Instance file systems are isolated; it is not possible for sandboxed code to access another instance's writeable directory.
```

MarkKoz reviewed (changes requested):
Just a couple comments!

MarkKoz reviewed (approved):
!!! excellent work

MarkKoz commented:
@wookie184 Do you still plan on continuing your review?

ionite34 commented:
Minor update in [3dfde52](https://github.com/python-discord/snekbox/pull/159/commits/3dfde52eda8f2af0475855c26bd5c1c2d2a0818d) to truncate `FileAttachment.__repr__` to avoid logs being spammed with megabytes of data from large user written files.

wookie184 reviewed (approved):
I don't think i'll have a chance to test this thoroughly, but this looks good.

MarkKoz commented on snekbox/nsjail.py:
```diff
@@ -163,37 +220,61 @@ def python3(
             returncode = -nsjail.returncode + 128 if nsjail.returncode < 0 else nsjail.returncode
 
+            # Exclude user uploaded files from output
+            to_exclude = [file.path for file in files]
```
Why send a string instead of a path object?

ionite34 replied:
is now a dictionary of Path and original last modified time floats in https://github.com/python-discord/snekbox/commit/36814de6f7a231c86b356ec27b649da5ffc85c44

MarkKoz commented on snekbox/nsjail.py:
```diff
@@ -163,37 +220,61 @@ def python3(
             # convert negative exit codes to the `N + 128` form.
             returncode = -nsjail.returncode + 128 if nsjail.returncode < 0 else nsjail.returncode
 
+            # Exclude user uploaded files from output
```
What if the files changed?

ionite34 replied:
updated to upload user files if last modified time changes.
in  https://github.com/python-discord/snekbox/commit/36814de6f7a231c86b356ec27b649da5ffc85c44,

MarkKoz commented:
Is there a reason we would rather implement a file ignore system than disable Python bytecode generation? Are there eval use cases that would rely on it being generated?

Does the current ignore system support ignoring directories? The fnmatch docs state that a forward slash is not treated specially, but I'm unsure of what that means in practice. In any case, it looks more limited than gitignore so documentation is better off avoiding making such comparison.

ionite34 commented:
> Is there a reason we would rather implement a file ignore system than disable Python bytecode generation? Are there eval use cases that would rely on it being generated?
> 
> Does the current ignore system support ignoring directories? The fnmatch docs state that a forward slash is not treated specially, but I'm unsure of what that means in practice. In any case, it looks more limited than gitignore so documentation is better off avoiding making such comparison.

it stemmed from this following code example which seems to generate a pyc file regardless of the PYTHONDONTWRITEBYTECODE env var on nsjail
```py
from pathlib import Path

Path("lib.py").write_text("func = lambda: 'hello'")

from lib import func

print(func())
```

MarkKoz commented:
Does using `python -B` fix it? https://docs.python.org/3/using/cmdline.html#cmdoption-B

ionite34 commented:
> Does using `python -B` fix it? [docs.python.org/3/using/cmdline.html#cmdoption-B](https://docs.python.org/3/using/cmdline.html#cmdoption-B)

Seems to, yes. I removed the file ignore filter and added the -B flag.

MarkKoz reviewed (approved):
Thanks, glad that fixed it
